### ISSUE TITLE
Automatically hide taskbar issues

### ISSUE THREAD

**Issue by shanyan-wcx** [torchgm/RoundedTB#36]
When the taskbar is automatically hidden, the application cannot be full screen, and there will be a process of turning into a fillet when the taskbar is displayed.

![屏幕截图 2021-09-05 130900](https://user-images.githubusercontent.com/58252651/132115988-b2760eeb-a7bd-4ac7-80fd-12ea44050b46.png)

**Comment by shanyan-wcx** [comment]
I found that the former seems to be a bug in win11 system.

**Comment by torchgm** [comment]
Auto-hide isn't completely supported, although it appears you have some additional taskbar mods that may be interfering. Please let me know what they are so I can look into this issue further.

**Comment by Cyberdroid1** [comment]
I just tried out the auto-hide taskbar feature while RoundedTB is running. 

I observed that when RTB is running, I cannot hover near the bottom edge to bring up the taskbar. Hovering would not bring it up. Pressing Start button or pressing Win+T would still bring them up. 

As soon as I closed RTB, it started working correctly again. After running RTB again, taskbar stopped opening on hover again.

**Comment by Cyberdroid1** [comment]
I am on the latest version of RoundedTB, and I noticed that after having autohiding taskbar enabled, the taskbar would flicker when bringing the mouse to the bottom, with the app crashing after a while.

**Comment by shanyan-wcx** [comment]
I used trafficmonitor. Can you be compatible with it in the future?

**Comment by shanyan-wcx** [comment]
I used trafficmonitor. Can you be compatible with it in the future?

**Comment by Cyberdroid1** [comment]
@shanyan-wcx your request is not connected to this issue. It would be better to either look if there is already an issue or open one yourself :)

**Comment by enoversum** [comment]
I get the same flickering of the whole-length normal taskbar on Windows 11 (Beta) with the current version of RoundedTB when I move my mouse towards the edge of the screen as @Cyberdroid1 reported. 
When I move my mouse towards the auto-hidden taskbar, first the whole taskbar appears, then only your mod parts are showing (only docked icons in the middle, or additional systray on the right). No other taskbar mods running. Also, there is a constant flickering (a few pixels high) along the whole width at the bottom as soon as RoundedTB is running.

**Comment by torchgm** [comment]
Yeah, turns out Windows' autohide feature directly interferes with RoundedTB as it also tries to set regions. I've given up supporting it.

However, I am working on an alternative autohide mode as part of RoundedTB that should entirely mitigate any flickering issues, be a lot snappier and also have a toggleable animation. No word on when exactly it'll be implemented or if it will even be feasible, but from my initial tests it seems to be something that I can hopefully have ready for the next couple of updates.

**Comment by enoversum** [comment]
> Yeah, turns out Windows' autohide feature directly interferes with RoundedTB as it also tries to set regions. I've given up supporting it.
> 
> However, I am working on an alternative autohide mode as part of RoundedTB that should entirely mitigate any flickering issues, be a lot snappier and also have a toggleable animation. No word on when exactly it'll be implemented or if it will even be feasible, but from my initial tests it seems to be something that I can hopefully have ready for the next couple of updates.

@torchgm Too bad that this is the case so far. Classic MacOS-style docks often hide the taskbar completely and implement their very own (working) version of auto-hiding docks, so I assume there must be ways of achieving this. 

Thanks for letting us know you'll be working on some kind of version of that! I understand this takes time. I for one wouldn't want to live without an auto-hiding taskbar, as I barely use it when I actively work in an app for hours :) .

**Comment by torchgm** [comment]
Yeah, most other docks are just windows with their own completely custom animations/triggers. Because it's the process' own window, it has a lot more control in how it customises it.

Meanwhile, the taskbar is a special type of window called an [Application Desktop Toolbar](https://docs.microsoft.com/en-us/windows/win32/shell/application-desktop-toolbars), or AppBar. AppBars are pretty cool in that they can automatically snap to monitor edges, resize the work area (the space maximised windows take up), and stack on top of each other.

They can also autohide, which they do by just actually moving the AppBar across the screen (which is also why the autohide animation can be so inconsistent - it depends on how busy DWM is). Problem is it can then run onto other monitors and overhang, so they use a region to prevent that. All of this applies to the taskbar as well, which uses the same autohide behaviour. As a result of it constantly setting a region as it moves, it resets the rounded/split effects that RTB applies. This means the taskbar either becomes solid again as it slides down or it flickers significantly as it moves.

Luckily, Windows lets programs specify the work area themselves, and also alter the visibility, opacity etc. of windows, which means I can implement my own autohide. This was a very early experimental attempt just to see if it'd be possible, it'll obviously be more fully-fleshed out and customisable by the time it makes it into an update:

https://user-images.githubusercontent.com/31840547/136594644-0ddb5d8b-dbd3-4bc7-a78e-63fdcb52fa37.mp4

**Comment by torchgm** [comment]
Animations are hard with how limited customisation of the taskbar is, but the current plan is for an (optional) fade-in/out.

**Comment by enoversum** [comment]
Thanks for letting me know 🙂!

**Comment by szeredniklaszlo** [comment]
Can we get a compiled version with this included? :D This is awesome!

**Comment by XXVtc** [comment]
Hey man, I would love to test this feature already.
Is there any way for interested people to test it out for now?

**Comment by torchgm** [comment]
Not yet - I've actually stripped out all the code for it. It was in some Canary build from a while ago but I'm focusing on finalising R3.1 first as that fixes a lot of other minor issues that cause confusion.

**Comment by crutchcorn** [comment]
Please be patient with @torchgm regarding this feature and others. There are many many reasons a developer may not want to share RC/beta/alpha code or builds.

Torch, I wish I could contribute more meaningfully via code, but unfortunately do not have much experience with native code

**Comment by Peedee2002** [comment]
Thanks for working on this issue - honestly even windows's implementation of autohide sucks (some apps will not let you see the taskbar when you move cursor to bottom, and the popups for stuff like language settings and start are built without the autohide feature in mind). A fade-in fade-out sounds more promising than the pop-in pop-out that exists.

**Comment by stephtr** [comment]
That looks awesome!
An additional feature I would be really interested in would be to show the taskbar as long as it isn't (partially) covered by another window and automatically hide it as soon as there's an overlap, similar to some docks on Linux.
I guess that's a feature which can be worked on once auto-hiding itself is finished 🙂

**Comment by bigQY** [comment]
> 

https://user-images.githubusercontent.com/52437374/146666970-b047e361-9179-426f-ae37-64ea22352d9a.mp4

The solution of this project(https://github.com/ChanpleCai/SmartTaskbar) may be helpful to you

But when my mouse pointer moves to the bottom of the screen, the taskbar still doesn't appear normally.This is the problem caused by RoundedTB

**Comment by stephtr** [comment]
Ah, I actually used that app a few years ago. Unfortunately, their approach doesn't work well with multiple monitors.
I'm already curious to see @torchgm's code 😁

**Comment by TechStudent11** [comment]
> Animations are hard with how limited customisation of the taskbar is, but the current plan is for an (optional) fade-in/out.

So when are we gonna get something like this in a canary/RC build: https://twitter.com/TorchCTI/status/1478450406837653509?cxt=HHwWisC42fqzwYQpAAAA?

**Comment by parazite7** [comment]
Yep that works perfectly, seems as though having a tiny sliver of the taskbar at the bottom of the screen visible even when "hidden" by setting the margin to 0 or 1 (so your mouse can actually overlap it) makes it work

**Comment by italoacr** [comment]
Apparently the flicking bug is gone 🤩

https://user-images.githubusercontent.com/89114441/182286056-effb7a41-67b5-4cd8-822e-2de37d751e3e.mp4

**Comment by enoversum** [comment]
Liebe Kundin, liebe Kollegin, lieber Kunde, lieber Kollege,

ich bin vom 30. Juli bis 5. August leider nicht im Büro und kann deshalb weder Deine/Ihre E-Mail lesen noch Dein/Ihr Anliegen bearbeiten. In wirklich dringenden Fällen bin ich telefonisch erreichbar. Aber auch nur, wenn die Hütte brennt. 

Nach meiner Rückkehr kümmere ich mich gern auch um Deine/Ihre nicht-dringenden Anliegen.


Mit den besten Grüßen

Eno Thiemann und der halbautomatische E-Mail-Beantworter

**Comment by Peedee2002** [comment]
> Apparently the flicking bug is gone 🤩
> Apparently.the.flicking.bug.is.gone.mp4

oooh what build is this? and what version of windows?

**Comment by italoacr** [comment]
> 
Hi :)
RoundedTB version 3.1

Windows 11 Home Single Language
Version 21H2
Compilação do SO	22000.795

**Comment by enoversum** [comment]
I can confirm, no more flickering. Only on Windows 11 22621.440 (Beta Preview), RoundedTB in Dynamic Mode isn't centred completely by default but needs a re-setting in the Windows Taskbar settings to be centred again. Other than that, consider me happy!

**Comment by FreeTurk** [comment]
for @torchgm, I have an idea for the auto-hide issue but I am not sure about it.

From what I have seen, what triggers autohide is a 1px tall border on top of the taskbar, which you can slightly see when the unmodified taskbar is hidden. If you cannot hover over that border, in a case such as roundedTB's margins hiding that border, you cannot make it appear. What can be done (as I said, if possible, I don't know it's feasibility on the code side) is to show that top border while the TB is hidden, and hide it back again when the TB is up.

_Or_ you can custom make some sort of pressure-activated auto-hide, which I would prefer. Honestly, f*ck the way Windows handles auto-hide where my slightest touch makes it appear. But for this one, I know it would be heavy for you, so don't even count it as a suggestion.

So what basically should be done is to drop the top margin to 0 while taskbar is hidden and back to whatever value it was when it is risen. I think it can be done since you mentioned the windows class of TB does the autohide by shifting the window (correct me if I am wrong) so you can watch the vertical position of the TB and change depending on that.

- Emir Öven

**Comment by edluksss** [comment]
for @torchgm and for everyone that wants to fix the auto-hide issue and start to use this amazing personalitation-tool. This bug can be fixed easily if you click on "..." and set each parameter to -0. This is going to solutionate your problems :)

**Comment by SOUMYAK2528** [comment]
taskbar not showing on hover... i want to fix this bug....please help

**Comment by Jannik02x** [comment]
> for @torchgm and for everyone that wants to fix the auto-hide issue and start to use this amazing personalitation-tool. This bug can be fixed easily if you click on "..." and set each parameter to -0. This is going to solutionate your problems :) This solution can make the task bar start blinking if you also have translucentTB and the compatibility option activated, you simply have to deactivate it and everything will work perfectly.

Thank you! This works perfectly for me. :D

**Comment by Dobby-progam** [comment]
I got the auto hide to work fine with no problems on windows 11 Home 22621.1413 22H2

**Comment by gloatoriginal** [comment]
https://user-images.githubusercontent.com/47373236/232345651-3c2ffee5-a691-4d26-b147-c4f452efe222.mp4

Currently receiving the flashing like from the OP post, Windows 11 Version 22H2, Build: 22624.1610 with the following settings:
![image](https://user-images.githubusercontent.com/47373236/232345690-a75eb7e8-4865-480d-a451-271dc48953af.png)

**Comment by Manas-P** [comment]
> When the taskbar is automatically hidden, the application cannot be full screen, and there will be a process of turning into a fillet when the taskbar is displayed.
> 
> ![屏幕截图 2021-09-05 130900](https://user-images.githubusercontent.com/58252651/132115988-b2760eeb-a7bd-4ac7-80fd-12ea44050b46.png)

Just update the margin to 0 or 1
![image](https://user-images.githubusercontent.com/62294926/234274105-9fd5a7fc-f3ef-4792-bcac-38690d53bbe0.png)

**Comment by Void48** [comment]
> Yeah, most other docks are just windows with their own completely custom animations/triggers. Because it's the process' own window, it has a lot more control in how it customises it.
> 
> Meanwhile, the taskbar is a special type of window called an [Application Desktop Toolbar](https://docs.microsoft.com/en-us/windows/win32/shell/application-desktop-toolbars), or AppBar. AppBars are pretty cool in that they can automatically snap to monitor edges, resize the work area (the space maximised windows take up), and stack on top of each other.
> 
> They can also autohide, which they do by just actually moving the AppBar across the screen (which is also why the autohide animation can be so inconsistent - it depends on how busy DWM is). Problem is it can then run onto other monitors and overhang, so they use a region to prevent that. All of this applies to the taskbar as well, which uses the same autohide behaviour. As a result of it constantly setting a region as it moves, it resets the rounded/split effects that RTB applies. This means the taskbar either becomes solid again as it slides down or it flickers significantly as it moves.
> 
> Luckily, Windows lets programs specify the work area themselves, and also alter the visibility, opacity etc. of windows, which means I can implement my own autohide. This was a very early experimental attempt just to see if it'd be possible, it'll obviously be more fully-fleshed out and customisable by the time it makes it into an update:
> 
>  AggravatingSoupyIvorybilledwoodpecker.mp4

Do you still plan on releasing this? The current default auto hide animation in Windows 11 is terrible
